core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java

```diff
@@ -72,7 +72,7 @@
   public void init(InitParameters params) {
     services = new EnumMap<>(CompactionKind.class);
 
-    var defaultService = CompactionDirectives.builder().build();
+    var defaultService = CompactionDirectives.builder().setService("default").build();
```

Comment:
```suggestion
    var defaultService = CompactionDirectives.builder().build();
```

Reply:
Or, could use the name of the `SimpleCompactionDispatcher`?

Reply:
The service name is used to get compaction service config and uniquely identify the service within the tserver.

Compaction service properties have the following prefix.
```
tserver.compaction.major.service.<service id>
```

With the default service currently having multiple default props w/ the following prefix.

```
tserver.compaction.major.service.default
```

Using the classname  would not result in very nice default props.

Reply:
I am fine with the name. I just thought it made more sense setting the value of "default" in the `SimpleCompactionDispatcher` class because this _is_ the default implementation. Your suggestion puts this logic into the builder, which I don't think should have any of the default business logic.

Reply:
I agree my name change suggestion doesn't make sense, and that `default` makes more sense here, since that's the name of the service that `SimpleCompactionDispatcher` is dispatching *to*. So, I agree with @milleruntime more than @keith-turner here, and think that setting it in `SimpleCompactionDispatcher` is more intuitive. I would rather the builder simply have it as a required parameter.

Also, I think a lot of this would be more clear if the Directive were called a Dispatch. Then, it would be more clear that what is being built by the builder is the responsibility of the Dispatcher, because it is the Dispatcher's job to set the name of the service that the Dispatch is being dispatched *to*.

Reply:
> Also, I think a lot of this would be more clear if the Directive were called a Dispatch. Then, it would be more clear that what is being built by the builder is the responsibility of the Dispatcher, because it is the Dispatcher's job to set the name of the service that the Dispatch is being dispatched to.

This seems reasonable to me. I agree with @keith-turner though that the name should be consistent with the `ScanDispatcher`, since they follow similar patterns. I didn't want to touch any of the Scan* classes in this PR though, as it was intended to clarify `CompactionDirectives` so I think renaming them both could be done separately.

Reply:
There are three additional cases to consider.

 * Future additions of fields to CompactionDirectives would require that the builder be able to default those or require the user to specify them breaking existing code written against the SPI. ScanDirectives currently has three fields and one can do ScanDirectives.builder().build() to create a directives with defaults for all three fields.
 * Users writing their own CompactionDispatcher who want to fall back to the default service would need to use the string `default` in their own code.
 * A user can write   CompactionDirectives.builder().build() and it would compile, but I think would result in an NPE at runtime.  If its going to be required, then it would probably be best to structure the code such that its impossible to produce code that compiles w/o specifying it.

The three cases lead me to support the user not having to specify the default when building a CompactionDirectives

Reply:
I see what you are saying that a call to `CompactionDirectives.builder().build()` should work properly and set the default values. Typically this is fine if the values set in there will work for most purposes. This could be my limited understanding of the new compaction code but I think the problem here is that the default values for `CompactionDirectives` are tied directly to the implementation in `SimpleCompactionDispatcher`. 

Yes, if I wrote my own CustomCompactionDispatcher, it should keep working with future updates to CompactionDirectives interface. But why would my CustomCompactionDispatcher call CompactionDirectives.builder().build() ? That would set the values for SimpleCompactionDispatcher. Wouldn't CustomCompactionDispatcher call CompactionDirectives.builder().setService("customCD").build() ?

Reply:
> his could be my limited understanding of the new compaction code but I think the problem here is that the default values for CompactionDirectives are tied directly to the implementation in SimpleCompactionDispatcher

The default values in this case are more tied to the default in Accumulo's Property.java

> But why would my CustomCompactionDispatcher call CompactionDirectives.builder().build() ?

If the javadoc states that one should not do that, then no one should.  Its just an API/SPI design principal that if an API/SPI can be designed such that user errors are caught at compile time instead of runtime, then compile time would be better if it can be done in a reasonable way.

Reply:
> API/SPI can be designed such that user errors are caught at compile time instead of runtime

For sure. I can do that in this PR.

Reply:
> The default values in this case are more tied to the default in Accumulo's Property.java

I had a question about this that perhaps you can clarify: am I using the terminology correctly if I say that there is a **major compaction service** named *default*, and that this **service** named *default*, by default, has 3 **executors** named *small*, *medium*, and *large*? If I have the terminology correct so far, is it correct to say that the only default value from the CompactionDirectives builder that ties to what's in Property.java is the name *default* to refer to the named **service**?

My understanding is that the CompactionDirectives is basically just assignment information to send the Directive (which I would call Dispatch) to the specified service. Is that correct?

If I am understanding all this correctly, then I think we should just make `setService()` a required method call before one is capable of calling `build()`. Since the responsibility of assigning the dispatch to a specific service is pretty much the only thing that the Dispatcher is responsible for doing, then I don't think it should rely on any defaults in the builder... but declare it explicitly.

Reply:
> that this service named default, by default, has 3 executors named small, medium, and large?

That is correct and the by the service has a planner configured in Property.java

> only default value from the CompactionDirectives builder that ties to what's in Property.java is the name default to refer to the named service?

That is correct.

> If I am understanding all this correctly, then I think we should just make setService() a required method call before one is capable of calling build() (possibly calling it toService()

If it's required at compile time then I would be ok w/ that change.  Also I agreep setService is not a good name for a fluent builder, something else would be better like toService().